File: u50_check/python/check50/bridge.py
"""The process bridge: discovery and invocation of Python checks.

The Rust runner spawns `python -m check50.bridge <cmd>` once to
discover the check registry (declaration order, descriptions,
dependencies, timeouts) and once per check to invoke it, passing the
dependency's return value via a JSON state file in the run dir.
"""

from __future__ import annotations

import importlib.util
import json
import os
import sys
import traceback
from typing import Any

from . import bridge_state as state
from .errors import Failure, Mismatch

STATE_FILE = ".check50-state"


def _load_module(path: str) -> Any:
    spec = importlib.util.spec_from_file_location("checks", path)
    if spec is None or spec.loader is None:
        raise ImportError(f"could not load the checks module from {path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules["checks"] = module
    spec.loader.exec_module(module)
    return module


def _envelope(ok: bool, **extra: Any) -> None:
    envelope: dict[str, Any] = {
        "ok": ok,
        "log": list(state.log_lines),
        "data": dict(state.payload),
    }
    envelope.update(extra)
    print(json.dumps(envelope), flush=True)


def _error_envelope(exc: Exception) -> None:
    _envelope(
        False,
        error={
            "type": type(exc).__name__,
            "value": str(exc),
            "traceback": traceback.format_exc().splitlines(),
        },
    )


def discover(path: str) -> None:
    try:
        _load_module(path)
    except Exception as exc:
        print(traceback.format_exc(), file=sys.stderr)
        raise SystemExit(f"could not load the checks module {path}: {exc}") from exc
    checks = [
        {
            "name": name,
            "description": state.checks[name].description,
            "dependency": state.checks[name].dependency,
            "timeout": state.checks[name].timeout,
            "hidden": state.checks[name].hidden,
        }
        for name in state.order
    ]
    print(json.dumps({"checks": checks}), flush=True)


def invoke(path: str, name: str, state_file: str) -> None:
    state.reset()
    try:
        _load_module(path)
    except Exception as exc:
        _error_envelope(exc)
        return
    check = state.checks.get(name)
    if check is None:
        _envelope(
            False,
            error={
                "type": "UnknownCheck",
                "value": f"no registered check named {name}",
                "traceback": [],
            },
        )
        return
    fn = check.fn
    state_info: Any = None
    has_state = state_file != "-" and os.path.exists(state_file)
    if has_state:
        try:
            with open(state_file) as handle:
                state_info = json.load(handle)
        except Exception as exc:
            _error_envelope(exc)
            return
    try:
        result = fn(state_info) if has_state else fn()
    except Failure as failure:
        cause: dict[str, Any] = {
            "rationale": failure.rationale,
            "help": failure.help,
        }
        if isinstance(failure, Mismatch):
            cause["expected"] = failure.expected
            cause["actual"] = failure.actual
        _envelope(False, cause=cause)
        return
    except Exception as exc:
        _error_envelope(exc)
        return
    if result is None:
        # check50 parity: a dependency that returns None passes nothing
        # (the README hello pattern has zero-arg dependents).
        _envelope(True)
        return
    try:
        with open(STATE_FILE, "w") as handle:
            json.dump(result, handle)
    except Exception as exc:
        _error_envelope(exc)
        return
    _envelope(True)


def main() -> None:
    command = sys.argv[1]
    if command == "discover":
        discover(sys.argv[2])
    elif command == "invoke":
        invoke(sys.argv[2], sys.argv[3], sys.argv[4])
    else:
        raise SystemExit(f"unknown bridge command: {command}")


if __name__ == "__main__":
    main()

File: u50_check/src/python/bridge.rs
//! The process bridge: discovery and per-check invocation of Python
//! checks on the provisioned `CPython` (Phase 0-2 of
//! `docs/U50_CHECK_PYTHON_PLAN.md`). Discovery runs the shipped package
//! once and emits the check registry as JSON; invocation spawns one
//! subprocess per check in its own process group (the runner's deadline
//! machinery kills the interpreter and every student process it spawned).

use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use anyhow::Context as _;
use serde::Deserialize;

use super::venv;
use crate::api::Failure;
use crate::result::ErrorInfo;

/// A discovered check (from the registry emitted by the shipped
/// package's discovery pass).
#[derive(Debug)]
pub struct PyCheckSpec {
    pub name: String,
    pub description: String,
    pub dependency: Option<String>,
    pub timeout: Option<Duration>,
    pub hidden: Option<String>,
}

/// Why a Python check did not pass: either the check raised a
/// `check50.Failure` (the normal failure path) or something went wrong
/// inside the check process itself (rendered as `Cause::Error`, like
/// the panic path).
pub(crate) enum PyFailure {
    Check(Failure),
    Internal(ErrorInfo),
}

#[derive(Debug, Deserialize)]
struct Discovery {
    checks: Vec<DiscoveredCheck>,
}

#[derive(Debug, Deserialize)]
struct DiscoveredCheck {
    name: String,
    description: String,
    dependency: Option<String>,
    timeout: Option<f64>,
    hidden: Option<String>,
}

#[derive(Debug, Deserialize)]
struct Envelope {
    ok: bool,
    #[serde(default)]
    log: Vec<String>,
    #[serde(default)]
    data: serde_json::Map<String, serde_json::Value>,
    #[serde(default)]
    cause: Option<EnvelopeCause>,
    #[serde(default)]
    error: Option<ErrorInfo>,
}

#[derive(Debug, Deserialize)]
struct EnvelopeCause {
    rationale: String,
    #[serde(default)]
    help: Option<String>,
    #[serde(default)]
    expected: Option<String>,
    #[serde(default)]
    actual: Option<String>,
}

/// The per-check state file: each check's pickled return value lives in
/// its run dir; dependents inherit it through the run-dir copy (each
/// check overwrites it with its own state, so a dependent always reads
/// its direct dependency's value).
pub(crate) const STATE_FILE: &str = ".check50-state";

/// Discovers the check registry of a Python checks file (declaration
/// order, descriptions, dependencies, timeouts) by importing it on the
/// provisioned interpreter.
///
/// # Errors
/// Returns an error when provisioning, spawning, or parsing fails.
pub fn discover(checks_file: &Path) -> anyhow::Result<Vec<PyCheckSpec>> {
    let python = venv::interpreter()?;
    let out = Command::new(python)
        .args([
            "-m",
            "check50.bridge",
            "discover",
            &checks_file.display().to_string(),
        ])
        .env(
            "CHECK50_CHECK_DIR",
            checks_file
                .parent()
                .map_or_else(|| PathBuf::from("."), Path::to_path_buf),
        )
        .output()
        .context("run the check discovery pass")?;
    anyhow::ensure!(
        out.status.success(),
        "check discovery failed: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    let stdout = String::from_utf8_lossy(&out.stdout);
    let discovery: Discovery =
        serde_json::from_str(stdout.trim()).context("parse the discovery envelope")?;
    Ok(discovery
        .checks
        .into_iter()
        .map(|check| PyCheckSpec {
            timeout: check.timeout.map(Duration::from_secs_f64),
            name: check.name,
            description: check.description,
            dependency: check.dependency,
            hidden: check.hidden,
        })
        .collect())
}

/// Invokes one check: spawns the interpreter in the check's run dir
/// (own process group), feeds it the dependency's pickled state, and
/// translates the result envelope into the engine's failure model. Log
/// lines and payload data flow through the check context like they do
/// for native checks.
///
/// # Errors
/// Returns [`PyFailure::Check`] when the check raised a
/// `check50.Failure`, and [`PyFailure::Internal`] when the check
/// process itself failed (bad envelope, import error, ...). A check
/// that overruns `timeout` is killed (process group) and reports the
/// timeout failure.
pub(crate) fn invoke(
    ctx: &mut crate::api::CheckContext,
    checks_file: &Path,
    check: &str,
    timeout: Duration,
) -> Result<(), PyFailure> {
    let python = venv::interpreter().map_err(|error| internal_error(&error))?;
    let state_file = ctx.run_dir.join(STATE_FILE);
    let mut command = Command::new(python);
    command
        .args([
            "-m",
            "check50.bridge",
            "invoke",
            &checks_file.display().to_string(),
            check,
            &state_file.display().to_string(),
        ])
        .current_dir(&ctx.run_dir)
        .env(
            "CHECK50_CHECK_DIR",
            checks_file
                .parent()
                .map_or_else(|| PathBuf::from("."), Path::to_path_buf),
        )
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt as _;
        command.process_group(0);
    }
    let mut child = command.spawn().map_err(|e| {
        internal_error(&anyhow::anyhow!("could not run the check interpreter: {e}"))
    })?;
    let stdout = child.stdout.take();
    let stderr = child.stderr.take();
    let buffer: Arc<Mutex<String>> = Arc::default();
    if let Some(stdout) = stdout {
        let buffer = Arc::clone(&buffer);
        std::thread::spawn(move || {
            let mut text = String::new();
            let _ = std::io::Read::read_to_string(&mut { stdout }, &mut text);
            *buffer
                .lock()
                .unwrap_or_else(std::sync::PoisonError::into_inner) = text;
        });
    }
    // Drain stderr (import errors and tracebacks surface in the error
    // path when no envelope arrives).
    if let Some(stderr) = stderr {
        std::thread::spawn(move || {
            let mut sink = String::new();
            let _ = std::io::Read::read_to_string(&mut { stderr }, &mut sink);
        });
    }
    let deadline = Instant::now() + timeout;
    loop {
        if child.try_wait().is_ok_and(|status| status.is_some()) {
            break;
        }
        if Instant::now() >= deadline {
            crate::api::kill_child(&mut child);
            let _ = child.wait();
            return Err(PyFailure::Check(Failure::new(format!(
                "check timed out after {} seconds",
                timeout.as_secs()
            ))));
        }
        std::thread::sleep(Duration::from_millis(25));
    }
    let envelope = buffer
        .lock()
        .unwrap_or_else(std::sync::PoisonError::into_inner)
        .clone();
    let envelope: Envelope = serde_json::from_str(envelope.trim())
        .map_err(|e| internal_error(&anyhow::anyhow!("bad check envelope: {e}")))?;
    for line in envelope.log {
        ctx.log(line);
    }
    for (key, value) in envelope.data {
        ctx.data(key, value);
    }
    match (envelope.ok, envelope.cause, envelope.error) {
        (true, _, _) => Ok(()),
        (false, Some(cause), _) => {
            let failure = Failure {
                rationale: cause.rationale,
                help: cause.help,
                expected: cause.expected,
                actual: cause.actual,
            };
            Err(PyFailure::Check(failure))
        }
        (false, None, Some(error)) => Err(PyFailure::Internal(ErrorInfo {
            kind: error.kind,
            value: error.value,
            traceback: error.traceback,
            data: error.data,
        })),
        (false, None, None) => Err(internal_error(&anyhow::anyhow!(
            "the check process reported failure without a cause"
        ))),
    }
}

/// An internal (non-`Failure`) failure from the check process.
fn internal_error(error: &anyhow::Error) -> PyFailure {
    PyFailure::Internal(ErrorInfo {
        kind: "Error".to_owned(),
        value: format!("{error:#}"),
        traceback: vec![format!("Error: {}", error.root_cause())],
        data: serde_json::Map::new(),
    })
}

// Silence the unused-path warning on platforms without process groups:
// the kill path still runs there, just without group semantics.
#[allow(unused)]
fn _path_typecheck(path: &Path) -> PathBuf {
    path.to_path_buf()
}
